src/util/stop.c

```diff
@@ -29,6 +29,15 @@
 
 /* utility routines to implement the various stopping criteria */
 
+static double vector_norm(unsigned n, double *vec, double *w)
+{
+    unsigned i;
+    double ret;
+    for (i = 0; i < n; i++)
+        ret += (w ? w[i] : 1) * fabs(vec[i]);   /* this is L1 norm */
```

Comment:
rather than have this conditional inside the loop, might be better to just have two loops
```c
if (w)
    for (i = 0; i < n; i++) ret += fabs(vec[i]);
else
    for (i = 0; i < n; i++) ret += w[i] * fabs(vec[i]);
```